qiskit/providers/__init__.py

```diff
@@ -110,6 +109,394 @@
+
+If you have a quantum device or simulator that you would like to integrate with
+Qiskit you will need to write a provider. A provider will provide Terra with a
+method to get available :class:`~qiskit.provider.BackendV1` objects. The
+:class:`~qiskit.provider.BackendV1` object provides both information describing
+a backend and its operation for the :mod:`~qiskit.transpiler` so that circuits
+can be compiled to something that is optimized and can execute on the
+backend. It also provides the :meth:`~qiskit.providers.BackendV1.run` method which can
+run the :class:`~qiskit.circuit.QuantumCircuit` objects and/or
+:class:`~qiskit.pulse.Schedule` objects. This enables users and other Qiskit
+APIs, such as :func:`~qiskit.execute.execute` and higher level algorithms in
+Qiskit Aqua, to get results from executing circuits on devices in a standard
+fashion regardless of how the Backend is implemented. At a high level the basic
+steps for writing a provider are:
+
+ * Implement a :class:`~qiskit.providers.ProviderV1` subclass that handles
+   access to the backend(s).
+ * Implement a :class:`~qiskit.providers.BackendV1` subclass and its
+   :meth:`~~qiskit.providers.BackendV1.run` method.
+
+   * Add any custom gates for the backend's basis to the session
+     :class:`~qiskit.circuit.EquivalenceLibrary` instance.
+
+ * Implement a :class:`~qiskit.providers.JobV1` subclass that handles
+   interacting with a running job.
+
+For a simple example of a provider the
+`qiskit-aqt-provider <https://github.com/qiskit-community/qiskit-aqt-provider>`__
+package is worth using as an example.
+
+Provider
+========
+
+A provider class serves a single purpose: to get backend objects that enable
+executing circuits on a device or simulator. The expectation is that any
+required credentials and/or authentication will be handled in the initialization
+of a provider object. The provider object will then provide a list of backends,
+and methods to filter and acquire backends (using the provided credentials if
+required). An example provider class looks like::
+
+    from qiskit.providers import ProviderV1 as Provider
+    from qiskit.providers.providerutils import filter_backends
+
+    from .backend import MyBackend
+
+    class MyProvider(Provider):
+
+        def __init__(self, token=None):
+            super().__init__()
+            self.token = token
+            self.backends = [MyBackend(provider=self)]
+
+        def backends(self, name=None, **kwargs):
+            if name:
+                backends = [
+                    backend for backend in backends if backend.name() == name]
+            return filter_backends(backends, filters=filters, **kwargs)
+
+The key aspects to ensure are present is that any necessary information for
+authentication (if required) are present in the class and that the backends
+method matches the abstract interface and will return initialized Backend
+objects. The rest is up to the specific provider on how to implement.
```

Comment:
```suggestion
authentication (if required) are present in the class and that the backends
method matches the required interface. The rest is up to the specific provider on how to implement.
```
details about the backends method (returning initialized objects) could be written as a docstring in the mini class example